stub/src/main/java/io/grpc/stub/BlockingClientCall.java

```diff
@@ -0,0 +1,407 @@
+/*
+ * Copyright 2023 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.stub;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import io.grpc.ClientCall;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.stub.ClientCalls.StartableListener;
+import io.grpc.stub.ClientCalls.ThreadlessExecutor;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Represents a bidirectional streaming call from a client.  Allows in a blocking manner, sending
+ * over the stream and receiving from the stream.  Also supports terminating the call.
+ * Wraps a ClientCall and converts from async communication to the public sync paradigm.
+ *
+ * @param <ReqT> Type of the Request Message
+ * @param <RespT> Type of the Response Message
+ */
+public final class BlockingClientCall<ReqT, RespT> {
+
+  private static final Logger logger = Logger.getLogger(BlockingClientCall.class.getName());
+
+  private final BlockingQueue<RespT> buffer;
+  private final StartableListener<RespT> listener;
+  private final ClientCall<ReqT, RespT> call;
+
+  private final ThreadlessExecutor executor;
+
+  private volatile boolean writeClosed;
+  private Status closedStatus;
+
+  BlockingClientCall(ClientCall<ReqT, RespT> call, ThreadlessExecutor executor) {
+    this.call = call;
+    this.executor = executor;
+    buffer = new ArrayBlockingQueue<>(1);
+    listener = new QueuingListener(buffer, call);
+  }
+
+  /**
+   * Wait if necessary for a value to be available from the server. If there is an available value
+   * return it immediately, if the stream is closed return a null. Otherwise, wait for a value to be
+   * available or the stream to be closed
+   *
+   * @return value from server or null if stream has been closed
+   * @throws io.grpc.StatusRuntimeException If the stream has closed in an error state
+   */
+  public RespT read() throws InterruptedException {
+    try {
+      return read(Integer.MAX_VALUE, TimeUnit.DAYS);
+    } catch (TimeoutException e) {
+      throw new RuntimeException(e); // should never happen
```

Comment:
`AssertionError` so you don't need the comment and and it is clear if it ever happens? Technically it could happen, but it's over 100 years (int max nanoseconds), so nobody living cares. go/java-practices/impossible

Reply:
Yes, you use DAYS here, but it is turned into long NANOSECONDS later which can only represent 292 years in the future so the extra time is lost. Same limitation as `System.nanoTime()`.

Reply:
Changed to AssertionError.